packages/dev/core/src/Maths/math.vector.ts

```diff
@@ -178,8 +178,8 @@ export class Vector2 {
      * @param otherVector defines the other vector
      * @returns a new Vector2
      */
-    public addVector3(otherVector: Vector3): Vector2 {
-        return new Vector2(this.x + otherVector.x, this.y + otherVector.y);
+    public addVector3(otherVector: Vector3): this {
+        return new (<any>this.constructor)(this.x + otherVector.x, this.y + otherVector.y);

```

Comment:
I like the generics approach, but I am not sure I am happy about `this.constructor` casted to `any` to get the types correctly.
I'll try playing around with it later, unless you have a suggestion with no `any` :-)

Reply:
How about something like this:

```javascript
export type Vector4Constructor<T extends Vector4> = new (x: number, y: number, z: number, w: number) => T;
//... now the vector 4 add function:
    /**
     * Returns a new Vector4 as the result of the addition of the current Vector4 and the given one.
     * @param otherVector the vector to add
     * @returns the resulting vector
     */
    public add(otherVector: DeepImmutable<Vector4>): this {
        return new (<Vector4Constructor<this>>this.constructor)(this.x + otherVector.x, this.y + otherVector.y, this.z + otherVector.z, this.w + otherVector.w);
    }
``

Reply:
@RaananW To have this work I tried to modify your proposal like this
`export type Vector2Constructor<T extends Vector2> = (new () => T) | ((new (x: number, y: number) => T));`
But this still gives me "Expected 2 arguments, but got 0." when I try to use it like this
`const result = new (<Vector2Constructor<T>>value1.constructor)();`
I guess one solution is to not use the default constructor or do you have a better idea? :)

Reply:
the variables should be optional, similar to the original constructor, i.e. `x?: number, y?: number`